### PR TITLE
Fix path expander II

### LINKED ISSUE #674: `LocalPathExpander` should match only `folder_paths` or `file_paths` depending on the spec.
I am trying to use the `LocalPathExpander` for the Clandinin project but I am encountering the following issue. I have something like the following structure:

```python
.
└── base_directory
    ├── a_simple_pattern_1
    │   └── a_simple_pattern_1.bin
    └── a_simple_pattern_2
        └── a_simple_pattern_2.bin


```

Then I call:

```python
from neuroconv.tools.path_expansion import LocalPathExpander

# Specify source data (note this assumes the files are arranged in the same way as in the example data)
source_data_spec = {
    "a_source": {
        "base_directory": base_directory,
        "folder_path": "a_simple_pattern_{session_id}",
    }
}

path_expander = LocalPathExpander()
metadata_list = path_expander.expand_paths(source_data_spec)

folder_paths = [metadata_match["source_data"]["a_source"]["folder_path"] for metadata_match in metadata_list]
folder_paths
```

My expectation would be to have only:
```python
['tmp/base_directory/a_simple_pattern_2',
 'tmp/base_directory/a_simple_pattern_1']
```
But instead I get:

```python
['tmp/base_directory/a_simple_pattern_2',
 'tmp/base_directory/a_simple_pattern_1',
 'tmp/base_directory/a_simple_pattern_2/a_simple_pattern_2.bin',
 'tmp/base_directory/a_simple_pattern_1/a_simple_pattern_1.bin']
```

That is, the `LocalPathExpander` is matching both `folder_paths` and `file_paths` when I am only requesting folder_paths. This makes it harder for me to use `LocalPathExpander` as a source of arguments for defining my interface as I am getting extra terms in the resulting dictionary that I need to filter after.

@bendichter @CodyCBakerPhD 
What do you think guys? is this a feature or is this a bug? I feel that if you specificy `folder_path` in the spec you should only get folder back and the same for file_paths. I made an implementation and more complete tests on https://github.com/catalystneuro/neuroconv/pull/675

### PR DESCRIPTION
**Pull request by h-mayorquin**
After #675 @bendichter [mentioned ](https://github.com/catalystneuro/neuroconv/pull/675#issuecomment-1840847452) that we are still missing the case of nested over matching:

> "a_simple_pattern_{session_id}" applied to "tmp/base_directory/a_simple_pattern_2/nested/" will still return session_id="2/nested/" when it should not match

and proposed a solution of excluding metadata entries that have file separators (slashes in posix or double front slashes on windows) from the matches. This PR implements this solution and amends the tests to reflect that.

This should fix #674

### CODE REVIEW

**Comment by bendichter** [comment]
@h-mayorquin thanks for working on this. Let me see if it's possible with regex

**Comment by bendichter** [comment]
oh, that's not easy with with parse library. hmm.

File: tests/test_minimal/test_tools/test_expand_paths.py [hunk 1]
```diff
@@ -13,11 +13,11 @@
...
     base_directory = Path(tmpdir)
 
-    sub_directory1 = base_directory / "a_simple_pattern_1"
-    sub_directory2 = base_directory / "a_simple_pattern_2"
+    sub_directory1 = base_directory / "subject1" / "a_simple_pattern_1"
+    sub_directory2 = base_directory / "subject2" / "a_simple_pattern_2"
 
-    sub_directory1.mkdir(exist_ok=True)
-    sub_directory2.mkdir(exist_ok=True)
+    sub_directory1.mkdir(exist_ok=True, parents=True)
```
Comment (bendichter) [inline tests/test_minimal/test_tools/test_expand_paths.py:19]: oh cool I didn't know about the parents arg

File: tests/test_minimal/test_tools/test_expand_paths.py [hunk 2]
```diff
@@ -28,38 +28,37 @@ def test_only_folder_match(tmpdir):
...
 
-    # Instantiate LocalPathExpander
-
     path_expander = LocalPathExpander()
     metadata_list = path_expander.expand_paths(source_data_spec)
     folder_paths = [metadata_match["source_data"]["a_source"]["folder_path"] for metadata_match in metadata_list]
 
-    expected = {str(sub_directory1), str(sub_directory2), str(sub_directory3)}
+    # Note that sub_directory3 is not included because it does not conform to the pattern
+    expected = {str(sub_directory1), str(sub_directory2)}
```
Comment (bendichter) [inline tests/test_minimal/test_tools/test_expand_paths.py:49]: can you also test that the metadata was extracted correctly?

Comment (h-mayorquin) [reply tests/test_minimal/test_tools/test_expand_paths.py:49]: Done. I also added a docstring to `extract_metadata`

**Comment by bendichter** [comment]
@h-mayorquin what do you think of this? I factored out the file creation logic so it can be used my multiple tests and I added a test case to ensure that the parser is not combining multiple nested directories into a single metadata field

File: tests/test_minimal/test_tools/test_expand_paths.py [hunk 3]
```diff
@@ -94,10 +105,13 @@ def test_only_file_match(tmpdir):
 
     path_expander = LocalPathExpander()
     matches_list = path_expander.expand_paths(source_data_spec)
-    file_paths = [match["source_data"]["a_source"]["file_path"] for match in matches_list]
+    file_paths = set(match["source_data"]["a_source"]["file_path"] for match in matches_list)
```
Comment (h-mayorquin) [inline tests/test_minimal/test_tools/test_expand_paths.py:108]: Minor point but the `file_paths` are transformed to set before the test.

**Review by h-mayorquin — COMMENTED** [review]
@bendichter This is OK to me. The code duplication across the two tests did bother me and I was thinking to use a module scoped pytest fixture to get rid of it. I was on the edge though because of the trade-off between your test being very explicit which is good and some code duplication which is bad. 

But if you think this makes more sense and I was already on the edge it seems tha the average favors a bit of abstraction. 

If you are Ok with this we should mege!

File: tests/test_minimal/test_tools/test_expand_paths.py [hunk 4]
```diff
@@ -2,96 +2,137 @@
...
+    matches_list = path_expander.expand_paths(source_data_spec)
+    file_paths = set(match["source_data"]["a_source"]["file_path"] for match in matches_list)
 
-    expected = {str(file1), str(file2), str(file3)}
+    # Note that file3 is not included because it does not conform to the pattern
+    expected = {
+        str(base_directory / "subject1" / "a_simple_pattern_1" / "a_simple_pattern_1.bin"),
+        str(base_directory / "subject2" / "a_simple_pattern_2" / "a_simple_pattern_2.bin"),
+    }
     assert set(file_paths) == expected
```
Comment (bendichter) [inline tests/test_minimal/test_tools/test_expand_paths.py:115]:
```suggestion
    assert file_paths == expected
```

**Review by bendichter — APPROVED** [review]

**Pull request merged**